Improve dynamics section of docs (#2271)
qutip/qutip: doc/guide/dynamics/dynamics-time.rst

:obj:`.QobjEvo` shares a lot of properties with the :obj:`.Qobj`.

+----------------+------------------+----------------------------------------+
| Property       | Attribute        | Description                            |
+================+==================+========================================+
| Dimensions     | ``Q.dims``       | List keeping track of shapes           |

> |                |                  | the tensor structure.                  |

Reviewer: `shapes in the tensor structure`.

> +================+==================+========================================+
> | Dimensions     | ``Q.dims``       | List keeping track of shapes           |
> |                |                  | the tensor structure.                  |

Author: ```suggestion
+================+==================+========================================+
| Dimensions     | ``Q.dims``       | Shapes the tensor structure.           |

```

Using arguments
---------------

> Until now, the coefficient were only functions of time. In the definition of ``H1_coeff``,

Reviewer: ```suggestion
Until now, the coefficients were only functions of time. In the definition of ``H1_coeff``,
```

or, new from v5, add the extra parameter directly:

.. code-block:: python

    >>> def H1_coeff(t, A, sigma):
    >>>     return A * np.exp(-(t / sigma)**2)

When the second positional input of the coefficient function is named ``args``,
the arguments are passed as a Python dictionary of ``key: value`` pairs.
Otherwise the coefficient function is called as ``coeff(t, **args)``.
In the last example, ``args = {'A': a, 'sigma': b}`` where ``a`` and ``b`` are the
two parameters for the amplitude and width, respectively.
This ``args`` dictionary need to be given at creation of the :obj:`.QobjEvo` when
function using then are included:

.. code-block:: python

    >>> system = [sigmaz(), [sigmax(), H1_coeff]]
    >>> args={'A': 9, 'sigma': 5}
    >>> qevo = QobjEvo(system, args=args)

But without ``args``, the :obj:`.QobjEvo` creation will fail:

.. code-block:: python

    >>> QobjEvo(system)
    TypeError: H1_coeff() missing 2 required positional arguments: 'A' and 'sigma'

When evaluation the :obj:`.QobjEvo` at a time, new arguments can be passed either
with the ``args`` dictionary positional arguments, or with specific keywords arguments:

> .. code-block:: python

Reviewer: It would be nice to keep these as runnable code if we can so we know the examples keep working.

Author: The plot directive only output plot...
Here the print command were there, but only the print, not the output...

Reviewer: Oh. I see. Maybe there is something we can add to the plot directive to make it output the console text too?

Author: I have not found any.
It would be great to have jupyter like cell in the documentation.

Reviewer: Perhaps let's leave these as is for now and just make a TODO somewhere to find a solution.

Lastly the spline method is usually as fast the string method, but it cannot be modified once created.

.. _time_max_step:

Working with pulses
===================

> Special care is needed when working with pulses. ODE solvers decide the step

Reviewer: ```suggestion
Special care is needed when working with pulses. ODE solvers select the step
```

length automatically and can miss thin pulses when not properly warned.

> Integrations methods with variables step have the ``max_step`` options that
> control the maximum length of a single internal integration step. This value
> should be set to under half the pulse width to be certain they are not missed.

Reviewer: ```suggestion
Integrations methods with variable step sizes have the ``max_step`` option that
control the maximum length of a single internal integration step. This value
should be set to under half the pulse width to be certain they are not missed.
```

For example, the following pulse is missed without fixing the maximum step length.

.. plot::
    :context: close-figs

    def pulse(t):
        return 10 * np.pi * (0.7 < t < 0.75)

    tlist = np.linspace(0, 1, 201)
    H = [sigmaz(), [sigmax(), pulse]]
    psi0 = basis(2,1)

    data1 = sesolve(H, psi0, tlist, e_ops=num(2)).expect[0]
    data2 = sesolve(H, psi0, tlist, e_ops=num(2), options={"max_step": 0.01}).expect[0]

    plt.plot(tlist, data1, label="no max_step")
    plt.plot(tlist, data2, label="fixed max_step")
    plt.fill_between(tlist, [pulse(t) for t in tlist], color="g", alpha=0.2, label="pulse")
    plt.ylim([-0.1, 1.1])
    plt.legend(loc="center left")

> In QuTiP 4.4 to 4.7, it was possible to request that the solver pass the state, expectation values or collapse operators via arguments to :obj:`.QobjEvo`. Support for this is not yet available in QuTiP 5.

Reviewer: Support is now available in QuTiP 5! :)

Perhaps just link to the relevant documentation from here.

Author: I meant to remove that line too.
The section has been moved to the next page.
